test-driver: Implement debugging breakpoint hooks

tfc opened the pull request:
This work picks up the core parts of https://github.com/NixOS/nixpkgs/pull/392117 (@Ma27)

I suppose that the review discussion might change some design aspects, so i would only document this at the end.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [Nixpkgs 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/doc/release-notes/rl-2511.section.md) (or backporting [25.05](https://github.com/NixOS/nixpkgs/blob/master/doc/manual/release-notes/rl-2505.section.md) Nixpkgs Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
- [NixOS 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2511.section.md) (or backporting [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) NixOS Release notes)
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md), [pkgs/README.md](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md), [maintainers/README.md](https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md) and other contributing documentation in corresponding paths.

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

tfc committed:
test-driver: Implement debugging breakpoint hooks

Co-authored-by: Maximilian Bosch <maximilian@mbosch.me>

tfc commented on nixos/lib/test-driver/default.nix:
```diff
@@ -25,13 +27,19 @@ python3Packages.buildPythonApplication {
 
+  postPatch = ''
+    substituteInPlace test_driver/debug.py \
+      --replace-fail "@attach@" "${breakpointHook.attach}"
```
instead of text-patching this, we could make the test driver parameter accept some kind of `--debug-hook ${breakpointHook.attach}` string and then use that in the `Debug` constructor. Not sure if overkill but not doing text patching might be nice and then it's also clear that the whole debug object is useless without this parameter.

Ma27 replied:
For me, the current approach was sufficient because I only iterated on this by `nix-build`ing the package (and was kinda "packager-brained" by patching sources when store-path references are needed).

So, while your solution is definitely cleaner, I don't see a strong reason to do so.

Hence, I'm fine with either option.

Ma27 commented on nixos/lib/test-driver/src/test_driver/driver.py:
```diff
@@ -224,8 +229,14 @@ def test_script(self) -> None:
                 for line in f"{exc_prefix}: {exc}".splitlines():
                     self.logger.log_test_error(line)
 
+                self.debug.break_on_failure()
```
So, I came to the conclusion that this isn't as useful as intended: when having to debug a failure, you already have to edit the config to enable the debugger and rerun. IMHO, that's not much benefit over setting a `breakpoint()` manually.

In fact, this has the benefit that you end up in code you "know" (i.e. your test script) rather than test driver internals (i.e. in one of the methods of the `Machine` class).

Ma27 reviewed (changes requested):
* Please use the standard GitHub way to attribute things in commits (i.e. a `Co-authored-by:` line in the commit).
* This is also missing the docs I wrote in https://github.com/NixOS/nixpkgs/pull/392117
* @mweinelt asked me to wait a bit with this: each change to the testdriver (and there were at least two in this week) costs a rebuild of all tests (including expensive ones like installers) meaning additional stress on the build infra. So it makes sense to wait with this for a few more days I guess.

Ma27 commented on nixos/doc/manual/development/writing-nixos-tests.section.md:
```diff
@@ -284,3 +284,54 @@ id-prefix: test-opt-
+  nodes.machine = {};
+
+  sshBackdoor.enable = true;
+  enableDenbugHook = true;
```
```suggestion
  enableDebugHook = true;
```

Ma27 reviewed (commented):
Looks good now.

Manual needs fixing as well before a merge.

Ma27 committed:
nixos/testing: enable ssh backdoor by default if debug hook is enabled

You usually want both, so it makes sense to have the former imply the
latter. If this is not desired, it can still be modified within a test.
